cardinal/config.go

```diff
@@ -160,3 +189,73 @@ func (w *WorldConfig) setLogger() error {
 
 	return nil
 }
+
+func setupViper() {
+	once.Do(func() {
+		// Bind the command-line flags to Viper
+		pflag.String(tomlConfigFileEnvVariable, "", "Path to the TOML config file")
+		pflag.Parse()
+		if err := viper.BindPFlags(pflag.CommandLine); err != nil {
+			log.Debug().Err(err).Msg("Failed to bind command-line flags to Viper")
+			// Continue even if the binding fails
+		}
+	})
+
+	// Bind the environment variables to Viper
+	viper.AutomaticEnv()
+
+	// Set default toml config file name and type
+	viper.SetConfigName("world") // name of config file (without extension)
+	viper.SetConfigType("toml")  // REQUIRED if the config file does not have the extension in the name
+
+	// Find the toml file from the flag and env variable
+	// viper precedence: flag > env > default
+	tomlFilePath := viper.GetString(tomlConfigFileEnvVariable)
+	if tomlFilePath != "" {
+		useSpecifiedTomlFile(tomlFilePath)
+	} else {
+		// Search for toml file in the current directory and parent directory
+		searchTomlInDirectories()
+	}
+
+	// Bind env from struct tags
+	val := reflect.ValueOf(&defaultConfig).Elem()
+	typ := val.Type()
+
+	for i := 0; i < val.NumField(); i++ {
+		field := typ.Field(i)
+		tag := field.Tag.Get("mapstructure")
+		if tag != "" {
+			if err := viper.BindEnv(tag); err != nil {
+				log.Warn().Err(err).Str("field", field.Name).Msg("Failed to bind env variable")
+			}
+		}
+	}
+}
+
+func searchTomlInDirectories() {
```

Comment:
imo, for the next 3 functions you can just inline them. 

this is too "clean code" that it ended up just being more confusing because the program flow jumps around.

Reply:
just got blamed by linter for nested if, so i'm separate it to the 3 func. 😁
i'll make another approach, or maybe i'll just use //nolint for this

Reply:
just use nolint if the linter is being stupid [](edited-with-Axolo)

Reply:
updated using nolint